packages/network/src/logic/node/Node.ts

```diff
@@ -255,11 +265,61 @@ export class Node extends EventEmitter {
     }
 
     async openOutgoingStreamConnection(streamPartId: StreamPartID, contactNodeId: string): Promise<void> {
-        await this.proxyStreamConnectionManager.openOutgoingStreamConnection(streamPartId, contactNodeId)
+        let resolveHandler: any
+        let rejectHandler: any
+        await Promise.all([
+            new Promise<void>((resolve, reject) => {
+                resolveHandler = (node: string, stream: StreamPartID) => {
+                    if (node === contactNodeId && stream === streamPartId) {
+                        resolve()
+                    }
+                }
+                rejectHandler = (node: string, stream: StreamPartID) => {
+                    if (node === contactNodeId && stream === streamPartId) {
+                        reject(`Joining stream as pure publisher failed on contact-node ${contactNodeId} for stream ${streamPartId}`)
+                    }
+                }
+                this.on(Event.PUBLISH_STREAM_ACCEPTED, resolveHandler)
+                this.on(Event.PUBLISH_STREAM_REJECTED, rejectHandler)
+            }),
+            this.proxyStreamConnectionManager.openOnewayStreamConnection(streamPartId, contactNodeId, Direction.OUT)
+        ]).finally(() => {
+            this.off(Event.PUBLISH_STREAM_ACCEPTED, resolveHandler)
+            this.off(Event.PUBLISH_STREAM_REJECTED, rejectHandler)
+        })
+    }
+
+    async closeOutgoingStreamConnection(streamPartId: StreamPartID, contactNodeId: string): Promise<void> {
+        await this.proxyStreamConnectionManager.closeOnewayStreamConnection(streamPartId, contactNodeId, Direction.OUT)
     }
 
-    closeOutgoingStreamConnection(streamPartId: StreamPartID, contactNodeId: string): void {
-        this.proxyStreamConnectionManager.closeOutgoingStreamConnection(streamPartId, contactNodeId)
+    async openInboundStreamConnection(streamPartId: StreamPartID, contactNodeId: string): Promise<void> {
+        let resolveHandler: any
+        let rejectHandler: any
+        await Promise.all([
+            new Promise<void>((resolve, reject) => {
+                resolveHandler = (node: string, stream: StreamPartID) => {
+                    if (node === contactNodeId && stream === streamPartId) {
+                        resolve()
+                    }
+                }
+                rejectHandler = (node: string, stream: StreamPartID) => {
+                    if (node === contactNodeId && stream === streamPartId) {
+                        reject(`Joining stream as pure subscriber failed on contact-node ${contactNodeId} for stream ${streamPartId}`)
+                    }
+                }
+                this.on(Event.SUBSCRIBE_STREAM_ACCEPTED, resolveHandler)
+                this.on(Event.SUBSCRIBE_STREAM_REJECTED, rejectHandler)
+            }),
+            this.proxyStreamConnectionManager.openOnewayStreamConnection(streamPartId, contactNodeId, Direction.IN)
+        ]).finally(() => {
```

Comment:
does this need some kind of system for removing these listeners in response to other events e.g. when node is disconnected or stream is unsubscribed?

Reply:
That's a good point, currently they get cleaned up by time outs at the latest. It would be useful to come up with specific scenarios where cleaning up the listeners fast would be required.

Currently if a proxy node gets disconnected, the ProxyStreamConnectionManager will attempt to reconnect to the node until `closeOnewayStreamConnection` is called. However, if the first attempt to create the proxy-connection fails `openOnewayStreamConnection` should throw and the listener is cleaned in the `finally` block.

In terms of a client unsubscribing in the current design, the client should explicitly tell its network-node to remove the proxy connection if it is no longer needed.